### PR TITLE
Hybrid mode support

### PR DESCRIPTION
**Pull request by jasonwaters**
## Description

This PR contains code that implements hybrid mode in alloy.  

I introduced a new command called `applyResponse`.  This command is similar to `sendEvent` in concept.  sendEvent constructs a konductor request and fulfills it client side.  But `applyResponse` accepts an object with two properties: `responseHeaders` and `responseBody`.  The value of these properties is a JSON object representing the headers or body of a konductor response.  The user is simply asking alloy to apply the already-fulfilled response.  By doing so any request to konductor could be made outside of alloy and then it's response applied to it.  This is not just for personalization -- any relevant handles will be applied (such as identity or state:store).

Details of Hybrid mode and the approach taken for this PR can be found on [this wiki page](https://wiki.corp.adobe.com/pages/viewpage.action?pageId=2615646420). 

Here is a samples repository that shows 4 different samples that illustrate the various ways you could implement personalization (server-side only, client-side only, hybrid and hybrid spa).  The hybrid samples use the new `applyResponse` command.  

view the repo ➡️  https://github.com/jasonwaters/aep-edge-samples

Here is an example of using the `applyResponse` command:

```js
alloy("applyResponse", {
  renderDecisions: true,
  responseHeaders: {
    "cache-control": "no-cache, no-store, max-age=0, no-transform, private",
    connection: "close",
    "content-encoding": "deflate",
    "content-type": "application/json;charset=utf-8",
    date: "Tue, 12 Jul 2022 16:02:48 GMT",
    server: "jag",
    "strict-transport-security": "max-age=31536000; includeSubDomains",
    "transfer-encoding": "chunked",
    vary: "Origin",
    "x-adobe-edge": "OR2;9",
    "x-content-type-options": "nosniff",
    "x-konductor": "22.6.78-BLACKOUTSERVERDOMAINS:7fa23f82",
    "x-rate-limit-remaining": "599",
    "x-request-id": "474ec8af-6326-4cb5-952a-4b7dc6be5749",
    "x-xss-protection": "1; mode=block",
  },
  responseBody: {
    requestId: "474ec8af-6326-4cb5-952a-4b7dc6be5749",
    handle: [
      {
        payload: [
          {
            id: "60808622677524487184147130113421486680",
            namespace: {
              code: "ECID",
            },
          },
        ],
        type: "identity:result",
      },
      {
        payload: [
          {
            id: "AT:eyJhY3Rpdml0eUlkIjoiNDU2NjE5IiwiZXhwZXJpZW5jZUlkIjoiMCJ9",
            scope: "sample-json-offer",
            scopeDetails: {
              decisionProvider: "TGT",
              activity: {
                id: "456619",
              },
              experience: {
                id: "0",
              },
              strategies: [
                {
                  step: "entry",
                  algorithmID: "0",
                  trafficType: "0",
                },
                {
                  step: "display",
                  algorithmID: "0",
                  trafficType: "0",
                },
              ],
              characteristics: {
                eventToken:
                  "YXv3o7b5b7uT6ivIpvEMCGqipfsIHvVzTQxHolz2IpSCnQ9Y9OaLL2gsdrWQTvE54PwSz67rmXWmSnkXpSSS2Q==",
              },
            },
            items: [
              {
                id: "1237147",
                schema:
                  "https://ns.adobe.com/personalization/json-content-item",
                meta: {
                  "geo.state": "utah",
                  "activity.id": "456619",
                  "experience.id": "0",
                  "activity.name": "aep-edge-samples sample-json-offer",
                  "offer.id": "1237147",
                  "profile.tntId":
                    "60808622677524487184147130113421486680-IesJBh",
                },
                data: {
                  id: "1237147",
                  format: "application/json",
                  content: {
                    heroImageName: "demo-marketing-offer1-exp-A.png",
                    buttonActions: [
                      {
                        id: 1,
                        text: "Buy Now",
                        content: "Thanks for your purchase!",
                      },
                      {
                        id: 2,
                        text: "Subscribe",
                        content: "You are now subscribed!",
                      },
                      {
                        id: 3,
                        text: "Download",
                        content: "Downloading...",
                      },
                    ],
                  },
                },
              },
            ],
          },
        ],
        type: "personalization:decisions",
        eventIndex: 0,
      },
      {
        payload: [
          {
            scope: "Target",
            hint: "35",
            ttlSeconds: 1800,
          },
          {
            scope: "AAM",
            hint: "9",
            ttlSeconds: 1800,
          },
          {
            scope: "EdgeNetwork",
            hint: "or2",
            ttlSeconds: 1800,
          },
        ],
        type: "locationHint:result",
      },
    ],
  },
}).then(applyMboxPersonalization("sample-json-offer"));
```

## Related Issue

https://jira.corp.adobe.com/browse/TNT-44491

## Motivation and Context

Staples and other target customers rely on hybrid mode in at.js today.  But they want to migrate to alloy/AEP.

### CODE REVIEW

**Comment by jasonwaters** [comment]
Link to prior PR for history: https://github.com/adobe/alloy/pull/863

**Comment by jasonwaters** [comment]
Hey @jonsnyder I am finished with the hybrid mode work after making several changes from the last team review:

- changed `applyHandles` to `applyAepEdgeResponse` which accepts `responseHeaders` and `responseBody` objects rather than just an array of handles.
- added additional unit test coverage
- added functional test that requests propositions from edge v2 api and applies them using `applyAepEdgeResponse`.  

I have also finished the samples and documentation for hybrid mode.  That effort can be found here: https://github.com/jasonwaters/aep-edge-samples/tree/main/target-offers-hybrid  . 

I consider hybrid mode feature complete at this point, so if you can review this PR and 👀 the related samples repo ☝️  that would be great.  thanks!

**Comment by ninaceban** [comment]
Thank you Jason for adding samples, this adds so much clarity. I'll review the functionality and add my review. 

I think we should be avoiding adding the branding words into our API. The reason behind this that we might rename the name of the brand, but the API is hard to be renamed.

File: src/core/edgeNetwork/injectApplyAepEdgeResponse.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    onResponseCallbackAggregator.add(lifecycle.onResponse);
+    onResponseCallbackAggregator.add(runOnResponseCallbacks);
+
+    return lifecycle
+      .onBeforeRequest({
+        request,
+        onResponse: onResponseCallbackAggregator.add,
+        onRequestFailure: noop
+      })
+      .then(() => {
```
Comment (jonsnyder) [inline src/core/edgeNetwork/injectApplyAepEdgeResponse.js:21]: Now that the customer is including the entire response, It may be useful to call processWarningsAndErrors like we do in sendEdgeNetworkRequest. This would catch a case where experience edge returns an error, but the customer doesn't know how to handle it. This also opens up the possibility that the "call" could fail in which case we should call the failure callbacks.

**Review by jonsnyder — CHANGES_REQUESTED** [review]
We talked in standup today about "applyNetworkResponse" for the name of the command. Do you like it?

**Comment by jasonwaters** [comment]
> 

Hi @jonsnyder , `applyNetworkResponse` is fine with me.  I don't have a strong opinion on the topic.  And I understand the desire to have a more generically named method name in case the marketing names change in the future.  

What i'd like to avoid is further name changes within this PR.  I have already renamed it several times during development.  On the surface, it doesn't seem like much effort, but it is actually quite a bit of work to rename.  Because I have to update the implementation code, sample code, documentation and flow diagrams.  So I'd prefer to change it one final time before merging.  Is the team settled on `applyNetworkResponse` as the name or do we need further discussion? 

@ninaceban

**Comment by jonsnyder** [comment]
Completely understand. We discussed it briefly in the standup yesterday. We could finalize the naming tomorrow morning at standup.

**Comment by jasonwaters** [comment]
Thank you @jonsnyder and @ninaceban for the feedback!

I have changed the command to be `applyResponse`as well as added error handling and several other changes based on your feedback.  The [samples](https://github.com/jasonwaters/aep-edge-samples) have also been updated, along with the documentation.

Please take another look when you have some time.

🍻

File: src/components/DataCollector/index.js [hunk 2]
```diff
@@ -65,6 +66,27 @@ const createDataCollector = ({ eventManager }) => {
             decisionScopes
           });
         }
+      },
+      applyResponse: {
+        documentationUri: "",
```
Comment (jfkhoury) [inline src/components/DataCollector/index.js:71]: Let's add a note to populate this field once documentation is ready.

**Review by jfkhoury — APPROVED** [review]
Awesome work on this feature @jasonwaters , including the requirement phase, samples, and ALLL the renaming exercises.

**Comment by jasonwaters** [comment]
@jonsnyder i can haz approval?

File: src/core/createEventManager.js [hunk 3]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable no-unused-vars */
```
Comment (jonsnyder) [inline src/core/createEventManager.js:1]: Why is this needed?

**Review by jonsnyder — APPROVED** [review]

File: src/core/createEventManager.js (hunk 3 above)
Comment (jasonwaters) [reply src/core/createEventManager.js]: it's not, thanks for catching it!
